### PR TITLE
Db store force triggered

### LINKED ISSUE #5020: Triggered task intercycle prereq bug
**Describe the bug**
A triggered task that doesn't get submitted before the workflow shuts down, gets stuck on the previous cycle prerequisite on restart, causing stall.

**Steps to reproduce the bug**
```ini
[scheduling]
    cycling mode = integer
    runahead limit = P1
    final cycle point = 5
    [[graph]]
        P1 = foo[-P1] => foo
[runtime]
    [[foo]]
        script = """
            sleep 2
            if [[ "$CYLC_TASK_CYCLE_POINT" == "4" ]]; then
                cylc stop "$CYLC_WORKFLOW_ID"
                cylc trigger "${CYLC_WORKFLOW_ID}//2/foo"
            fi
        """
```
First run log:
```
...
INFO - Command succeeded: stop(mode=Clean)
INFO - Processing 1 queued command(s)
INFO - [2/foo waiting(runahead) job:01 flows:1] spawned
INFO - [2/foo waiting(runahead) job:01 flows:1] => waiting
INFO - [2/foo waiting job:01 flows:1] => waiting(queued)    <-------------------------- 2/foo not yet submitted
INFO - Command succeeded: force_trigger_tasks(['2/foo'], flow=['all'], flow_wait=False, flow_descr=None)
INFO - [4/foo running job:01 flows:1] (received)succeeded at 2022-07-27T16:05:30+01:00
INFO - [4/foo running job:01 flows:1] => succeeded
INFO - [5/foo waiting(runahead) job:00 flows:1] spawned
INFO - [5/foo waiting(runahead) job:00 flows:1] => waiting
INFO - [5/foo waiting job:00 flows:1] => waiting(queued)
INFO - Workflow shutting down - REQUEST(CLEAN)
INFO - DONE
```
Restart log:
```
...
INFO - Flows:
    flow: 1 (original flow from 1) 2022-07-27 16:05:02
WARNING - Partially satisfied prerequisites:
      * 2/foo is waiting on ['1/foo:succeeded']    <-------------------------------
CRITICAL - Workflow stalled
```

**Expected behavior**
`2/foo` doesn't wait for `1/foo:succeed`, it should just run at restart because it was triggered.

**Screenshots**
(`2/foo` doesn't show up on restart)

image

**Workaround**

```
$ cylc set-outputs WORKFLOW_ID//1/foo
```

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

### PR DESCRIPTION
**Pull request by hjoliver**
These changes close #5020

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.
- [ ] [if bugfix] PRs raised to both master and the relevant bugfix branch.

### CODE REVIEW

**Comment by hjoliver** [comment]
(If we want to get this in the release, it is probably doable ... back in a couple of hours...)

**Review by MetRonnie — COMMENTED** [review]
Fixes the bug, just a couple of queries

File: CHANGES.md [hunk 1]
```diff
@@ -28,6 +28,17 @@ creating a new release entry be sure to copy & paste the span tag with the
 `actions:bind` attribute, which is used by a regex to find the text to be
 updated. Only the first match gets replaced, so it's fine to leave the old
 ones in. -->
+-------------------------------------------------------------------------------
+## __cylc-8.0.1 (<span actions:bind='release-date'>Upcoming</span>)__
+
+### Enhancements
+
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply CHANGES.md:35]: True dat.

**Comment by MetRonnie** [comment]
Sorry, accidentally deleted 8.0.x!

**Review by MetRonnie — COMMENTED** [review]
Looks good pending resolution of my flow nums query

**Comment by hjoliver** [comment]
Yes, I need to (re)-investigate that - I ~think I did it deliberately, and it seems to work correctly, but have forgotten my own chain of thought already...

File: cylc/flow/workflow_db_mgr.py [hunk 2]
```diff
@@ -441,7 +441,8 @@ def put_update_task_state(self, itask):
         set_args = {
             "time_updated": itask.state.time_updated,
             "status": itask.state.status,
-            "flow_wait": itask.flow_wait
+            "flow_wait": itask.flow_wait,
+            "is_manual_submit": itask.is_manual_submit
         }
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/workflow_db_mgr.py:443]: OK, turns out no DB insert or update is needed there at all. But worth investigating as I identified and remove a couple of redundant DB updates.

**Comment by MetRonnie** [comment]
I tried your example running on 8.0.x and this branch, but I can't say I noticed a difference in the changes to the `task_states` table

**Comment by hjoliver** [comment]
Sorry, poorly explained by me! - that was just intended to show that my changes to the task status DB calls had not broken anything.

**Review by MetRonnie — APPROVED** [review]
Looks reasonable to me. I'm out of my depth when it comes to knowing what some of the changes will mean, though

File: cylc/flow/task_job_mgr.py [hunk 3]
```diff
@@ -238,7 +238,6 @@ def prep_submit_task_jobs(self, workflow, itasks, check_syntax=True):
                 itask.submit_num += 1
                 itask.state_reset(TASK_STATUS_PREPARING)
                 self.data_store_mgr.delta_task_state(itask)
-                self.workflow_db_mgr.put_update_task_state(itask)
```
Comment (oliver-sanders) [inline cylc/flow/task_job_mgr.py:241]: This means the DB won't pick up on the PREPARING state until later on?

I think a `state_reset` should always be followed by a DB update?

File: CHANGES.md [hunk 4]
```diff
@@ -43,6 +43,9 @@ would not obey the given order.
 on a top level dir containing run dir(s) will now remove that top level dir
 in addition to the run(s) (if there is nothing else inside it).
 
+[#5023](https://github.com/cylc/cylc-flow/pull/5023) - Tasks force-triggered
+after a shutdown was requested, should submit to run immediately on restart.
+
```
Comment (MetRonnie) [inline CHANGES.md:48]: Now need to move to 8.0.2

Comment (hjoliver) [reply CHANGES.md]: Done.

**Comment by hjoliver** [comment]
(TBD: resolve Oliver's question above.)

**Review by MetRonnie — APPROVED** [review]

File: CHANGES.md [hunk 5]
```diff
@@ -28,6 +28,17 @@ creating a new release entry be sure to copy & paste the span tag with the
...
+-------------------------------------------------------------------------------
+## __cylc-8.0.2 (<span actions:bind='release-date'>Pending YYYY-MM-DD</span>)__
+
+Maintenance release.
+
+### Fixes
+
+[#5023](https://github.com/cylc/cylc-flow/pull/5023) - tasks force-triggered
+after a shutdown was ordered should submit to run immediately on restart.
```
Comment (MetRonnie) [inline CHANGES.md:40]: Now need to move to 8.0.3

Comment (hjoliver) [reply CHANGES.md:40]: Done.

File: cylc/flow/task_job_mgr.py (hunk 3 above)
Comment (hjoliver) [reply cylc/flow/task_job_mgr.py:241]: OK, I've gone back for a look at this.  The reason is, as per the comment added to the removed method:
```python
        This method is only needed as a final update for finished tasks,
        when they get removed from the task_pool.
```
The task pool and task states table is currently updated at the same time as the datastore, in the main loop via `Scheduler.update_data_structure()`

We have evidently been calling `put_update_task_state` in a couple of places in addition to that, but given the above that wasn't the intention.

We've been planning to all of these updates in more event-driven way, but that's beyond the scope of this PR and will required some discussion.  (Maybe should post a new issue for thoughts on this?).

Comment (oliver-sanders) [reply cylc/flow/task_job_mgr.py:241]: Ok, that makes sense.

**Review by MetRonnie — APPROVED** [review]

**Review by oliver-sanders — APPROVED** [review]

**Comment by MetRonnie** [comment]
Oops, this breaks the ability of Cylc to restart `8.0.{0,1,2}` workflows.
```
ERROR - no such column: task_states.is_manual_submit
    Traceback (most recent call last):
      File "~/github/cylc-flow/cylc/flow/scheduler.py", line 670, in start
        await self.configure()
      File "~/github/cylc-flow/cylc/flow/scheduler.py", line 459, in configure
        self._load_pool_from_db()
      File "~/github/cylc-flow/cylc/flow/scheduler.py", line 719, in _load_pool_from_db
        self.workflow_db_mgr.pri_dao.select_task_pool_for_restart(
      File "~/github/cylc-flow/cylc/flow/rundb.py", line 860, in select_task_pool_for_restart
        for row_idx, row in enumerate(self.connect().execute(stmt)):
    sqlite3.OperationalError: no such column: task_states.is_manual_submit
CRITICAL - Workflow shutting down - no such column: task_states.is_manual_submit
```
We need to make this an 8.1.0-only bugfix.
- We need to revert this change in the 8.0.x branch.
- It has already been merged into the master branch in #5158, but we need to move the changelog entry from the 8.0.3 section to 8.1.0.

**Comment by oliver-sanders** [comment]
Dammit, not sure how I missed that :facepalm:

**Comment by hjoliver** [comment]
Oops, sorry :face_with_spiral_eyes:

**Pull request merged**